plugins/scanners/fossid/src/main/kotlin/FossIdScanResults.kt

```diff
@@ -167,29 +169,40 @@ private fun mapLicense(
  */
 internal suspend fun mapSnippetFindings(
     rawResults: RawResults,
+    snippetsLimit: Int,
     issues: MutableList<Issue>,
     detectedLicenseMapping: Map<String, String>,
     snippetChoices: List<SnippetChoice>,
     snippetLicenseFindings: MutableSet<LicenseFinding>
 ): Set<SnippetFinding> {
     val remainingSnippetChoices = snippetChoices.toMutableList()
-    val allFindings = mutableSetOf<SnippetFinding>()
-
-    rawResults.listSnippets.map { (file, rawSnippets) ->
-        rawSnippets.mapSnippetFindingsForFile(
-            file,
-            rawResults.snippetMatchedLines,
-            issues,
-            detectedLicenseMapping,
-            snippetChoices,
-            remainingSnippetChoices,
-            snippetLicenseFindings
-        )
-    }.collect {
-        allFindings += it
+    val results = mutableSetOf<SnippetFinding>()
+    var runningSnippetCount = 0
+
+    coroutineScope {
+        launch {
+            rawResults.listSnippets.collect { (file, rawSnippets) ->
+                val mappedSnippets = rawSnippets.mapSnippetFindingsForFile(
+                    file,
+                    rawResults.snippetMatchedLines,
+                    issues,
+                    detectedLicenseMapping,
+                    snippetChoices,
+                    remainingSnippetChoices,
+                    snippetLicenseFindings
+                )
+
+                runningSnippetCount += mappedSnippets.size
+                results += mappedSnippets
+
+                if (runningSnippetCount >= snippetsLimit) {
```

Comment:
This `cancel()` looks rather drastic. Would it instead be possible to use `takeWhile` on the flow to enforce the limit?

Reply:
I thought about `takeWhile` too. I tried:
```
 rawResults.listSnippets.takeWhile { runningSnippetCount < snippetsLimit }.collect { (file, rawSnippets) ->
                val mappedSnippets = rawSnippets.mapSnippetFindingsForFile(
                    [...]
                )

                runningSnippetCount += mappedSnippets.size
                results += mappedSnippets
            }
        }
```

Unfortunately, when condition becomes `false`, the first element that must be ignored is nevertheless emitted to the flow.

fun <[T](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/take-while.html)> [Flow](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-flow/index.html)<[T](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/take-while.html)>.[takeWhile](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/take-while.html)(predicate: suspend ([T](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/take-while.html)) -> [Boolean](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-boolean/index.html)): [Flow](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-flow/index.html)<[T](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/take-while.html)>

This is because `T` must be emitted before the condition can be evaluated

Reply:
There is a more generic version called `transformWhile`. It allows more control over the elements that are emitted. I think, this could be used here.

Reply:
Ok, it works because it allows to evaluate the condition _after_ the emission.